Adopt base tag for derivative compute item

nilsvu opened the pull request:
## Proposed changes

This PR adds a compute item for the gradient of tensors in a Variables `ComputeDeriv`, similar to the existing `ComputeDiv`.

- [x] It is purely additive right now, and I'd like to get some feedback on the concepts, before removing the existing compute item functionality built into `Tags::deriv`. The change in the test shows that the new compute item works as expected.

The reason for this PR is mostly that I found the existing compute item functionality in `Tags::deriv` does not allow for prefixed variables because it takes a tags list instead of a Variables tag.

- [x] Depends on #900

### What breaks:

- Code that used `Tags::deriv` as a compute items needs to adopt the new `Tags::ComputeDeriv`. No changes are needed for code that used `Tags::deriv` as a simple item.

### Types of changes:

- [x] New feature

### Component:

- [x] Code

### Code review checklist

- [x] The PR passes all checks, including unit tests, `clang-tidy` and `IWYU`. For
  instructions on how to perform the CI checks locally refer to the [Dev guide
  on the Travis CI](https://spectre-code.org/travis_guide.html).
- [x] The code is documented and the documentation renders correctly. Run `make doc`
  to generate the documentation locally into `BUILD_DIR/docs/html`. Then open
  `index.html`.
- [x] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

nilsdeppe commented:
All this does is replicate existing functionality... Why does this PR exist?

nilsvu commented:
I wrote so in the description when I submitted the PR, as well as in Slack. If someone else wants to fix this, making `Tags::deriv` work for prefixed Variables, please do 😄
I named the tag `Tags::ComputeGrad` so that it corresponds to the existing `Tags::ComputeDiv`.

nilsdeppe commented:
Don't cite Slack, private email, private conversation, SpECTRE code meetings, etc.  The code is open-source, so any questions on PRs should be answered there.

Your response hasn't helped me. There is no test or anything demonstrating what the issue is and how you are fixing it.

nilsvu commented:
This is really not that complicated, and I expected the description in the PR to be enough 😄

Just look at the compute item version of `Tags::deriv`: it lists `Tags::Variables<variables_tags>` in its `argument_tags`. But if the `variables_tags` are in a prefixed Variables tags, e.g. `db::add_tag_prefix<SomePrefix, Tags::Variables<nonprefixed_variables_tags>>`, then the argument isn't found. This is what I fixed. And I named it `Tags::ComputeGrad` so that it corresponds to the existing `Tags::ComputeDiv` and doesn't break any code.

kidder commented:
@wthrowe and I agree that this file needs to be redone given the recent changes adding simple tags, ...
We like your new implementation of the compute item, just don't rename it.
As @nilsdeppe wants, please add a test case for the new usage that was not possible before.
We also think this PR should get rid of the deriv_impl functions as was done for Divergence as these are no longer necessary

nilsvu committed:
Adopt base tag for derivative compute item

The compute item is no longer a specialization of Tags::deriv, but instead
derives from it as Tags::ComputeDeriv. You can use Tags::deriv as a simple
item as you did before. The new compute item can handle prefixed variables.

nilsvu commented:
How about this?

I should note that I removed the compute item for `logical_partial_derivatives` because it was unused.

wthrowe commented on src/NumericalAlgorithms/LinearOperators/PartialDerivatives.hpp:
```diff
@@ -98,55 +102,41 @@ auto partial_derivatives(
+ * optionally specify a subset of the tags in the `Variables` as `GradientTags`
+ * to only take the gradient of those.
+ *
+ * This tag inherits from `db::add_prefix_tag<Tags::deriv, VariablesTag>`.
```
I don't think this inheritance is right.  Add a test that you can retrieve the computed Variables through the simple tag.

nilsvu replied:
Yeah something is wrong. Do you know how I can make the tag derive from the  `db::add_tag_prefix<Tags::deriv, VariablesTag>` with a `tags_list` that is restricted to the `GradientTags`? Keeping in mind that `VariablesTag` may carry prefixes already

wthrowe replied:
OK, yeah, that's a lot trickier than I thought it would be.  I'll submit a PR with a metafunction to set the tags_list of a Variables tag preserving prefixes shortly.  I think that along with things like `wrap_tags_in` should make this doable.

Edit: #900

nilsvu replied:
Worked nicely with the new function, thank you 👍

wthrowe reviewed (commented):
Looks good.  Squash.

nilsdeppe commented on src/NumericalAlgorithms/LinearOperators/PartialDerivatives.hpp:
```diff
@@ -98,55 +102,42 @@ auto partial_derivatives(
+ * optionally specify a subset of the tags in the Variables as `GradientTags` to
+ * only take the gradient of those.
+ *
+ * This tag inherits from `db::variables_tag_with_tags_list<VariablesTag,
```
Why do you need to document what it inherits from? Is it necessary to know this when I use the tag? Wouldn't it be more beneficial to document that you can use `deriv<Tag>` or whatever to grab the individual differentiated tensors from the databox?

wthrowe replied:
The base class is, in this case, an important part of the interface, because it tells you what tag you can retrieve the result of the compute item as (and for some reason doxygen does not show public base classes).  It implies that you can retrieve the tensors using `deriv<Tag>` because it shows that this is a variables tag and all variables tags work that way.

nilsdeppe replied:
The purpose should be made clear in the documentation, and why not document the other tags and how to retrieve them?

wthrowe replied:
Perhaps "may be retrieved via" or similar instead of "inherits from"?

I feel that we can assume some knowledge of how to use the DataBox when documenting DataBox tags.  We don't document how to retrieve tensors for all the other tags that obtain subitems through (possibly) containing Variables (`dt`, `deriv`, `ComputeDiv`, `NormalDotNumericalFlux`, ...), so we should come to some consistent decision on whether we want to put this everywhere.  At the moment common practice is to not do so.

nilsdeppe replied:
Yea, sorry for not being clear. I think using `May be retrieved via` would be perfect.

Good point. I think it would be sufficient to say what the appropriate base tag to use would be. So in this case saying something like `The variables may be retrieved via deriv<...>, while the tensors of the Variables may be retrieved using deriv<...>` (expanding on the ...) would be a good way to proceed and also should apply in general for any prefix and base tag combination. That should make it possible to easily figure out what the best way to retrieve an item is. What do you think?

wthrowe replied:
The idea of documenting subitem accesors is not unreasonable, but it is a fairly major change to a well-established documentation practice and I don't think we should be changing that practice based on discussion in a random comment thread in a bug about something else.  If we do want to do this, we should update all the other tags.  How about filing a bug so the discussion can get wider visibility and then assuming there is consensus for a change we can do it everywhere and make it all consistent?

nilsvu replied:
You have #739 open for this discussion already, let's use that

nilsdeppe commented on src/NumericalAlgorithms/LinearOperators/PartialDerivatives.hpp:
```diff
@@ -98,55 +102,42 @@ auto partial_derivatives(
+ *
+ * Computes the spatial derivatives of the Tensors in the Variables represented
+ * by `VariablesTag` in the frame mapped to by `InverseJacobianTag`. You can
+ * optionally specify a subset of the tags in the Variables as `GradientTags` to
```
How do I specify this subset? The way you've worded the documentation it sounds like I'm supposed to have those tags inherit off of a class call `GradientTags`, not specify a typelist.

nilsvu commented:
done

kidder reviewed (commented):
Looks good, squash into a single commit.  @wthrowe do you have any concerns?

kidder commented:
you have a documentation error to fix; the other failure is from #928

wthrowe commented on src/NumericalAlgorithms/LinearOperators/PartialDerivatives.hpp:
```diff
@@ -25,40 +25,44 @@ template <size_t Dim>
+                                                 UpLo::Lo, Frame>;
+  using tag = Tag;
+  static std::string name() noexcept { return "deriv(" + Tag::name() + ")"; }
+  static constexpr bool should_be_sliced_to_boundary = true;
```
The "should_be_sliced_to_boundary" variables were recently obsoleted (#921).  Remove them.

nilsvu commented:
done

kidder reviewed (approved):
@wthrowe this is ready too
